js-packages/@prestojs/rest/src/useAsyncValue.ts

```diff
@@ -0,0 +1,249 @@
+     */
+    error: null | Error;
+    /**
+     * True while `resolve` call is in progress.
+     */
+    isLoading: boolean;
+    /**
+     * The resolved value
+     */
+    value: T | null;
+    /**
+     * A function to manually trigger the action. If `options.trigger` is `MANUAL`
+     * calling this function is the only way to trigger the action.
+     *
+     * This function will return a promise that resolves/rejects to same value
+     * returned by `execute`. If `accumulatePages` is set the value returned is
+     * the accumulated value.
+     */
+    run: (...args) => Promise<any>;
+    /**
+     * When called will unset result and error to null. Will not immediately trigger
+     * a call to the action but subsequent changes to query or paginator will according
+     * to the value of `trigger`.
+     *
+     * If value is found in `existingValues` then it will still be returned even if
+     * you call `reset` - `reset` only clears values that are returned from `resolve`.
+     */
+    reset: () => void;
+};
+
+/**
+ * Resolve a value from an id using an async function.
+ *
+ * For the specified `id` the `resolve` function will be called and should
+ * return the value for the specified id.
+ *
+ * If `existingValues` is provided this will be consulted first before calling
+ * `resolve`. This is useful when you have a list of values that may or
+ * may not have the data you care about. If it's there then the value will
+ * be returned immediately.
+ *
+ * For multiple values see documentation below.
+ *
+ * @extract-docs
+ */
+export default function useAsyncValue<T, U extends Id>(
+    props: UseAsyncValuePropsSingle<T, U>
+): UseAsyncValueReturn<T>;
+/**
+ * Resolve values from an array of ids using an async function.
+ *
+ * For the specified array of `ids` the `resolve` function will be called
+ * and should return an array of the same size with each entry being the resolved
+ * value for the corresponding id.
+ *
+ * If `existingValues` is provided this will be consulted first before calling
+ * `resolve`. This is useful when you have a list of values that may or
+ * may not have the data you care about. If it's there then the values will
+ * be returned immediately. Note that if any of the ids are missing from
+ * `existingValues` then it will be ignored and a call to `resolve` will be made
+ * requesting values for all `ids`.
+ */
+export default function useAsyncValue<T, U extends Id>(
+    props: UseAsyncValuePropsMulti<T, U>
+): UseAsyncValueReturn<T[]>;
+export default function useAsyncValue<T, U extends Id>(
+    props: CommonProps<T, U> & {
+        id?: U | null;
+        ids?: U[] | null;
+        resolve: (idOrIds: U | U[]) => Promise<T[] | T>;
+    }
+): UseAsyncValueReturn<T | T[]> {
+    const { id, ids, getId, resolve, trigger } = props;
+    if (id && ids) {
+        throw new Error("Only one of 'id' and 'ids' should be provided");
```

Comment:
yeah typescript will pick this up because of overloads with `UseAsyncValuePropsSingle` and `UseAsyncValuePropsMulti` - just for runtime when not using TS